### PR TITLE
Spring Boot 2.7 compatibility and sync commits for tasks triggerer on partitions revoking.

### PR DESCRIPTION
**Pull request by onukristo**
## Context

* Tasks' triggers' offset is committed synchronously, when partitions are revoked.

* Reworked paranoid tasks cleaner to work with latest mariadb drivers.

* Made it compatible with Spring Boot 2.7

* Removing support for Spring Boot 2.4

## Checklist
- [x] Change meets or does not compromise the [Baseline Security Requirements](https://transferwise.atlassian.net/wiki/spaces/EKB/pages/434929973/Baseline+Security+Requirements)

### CODE REVIEW

File: tw-tasks-core/src/main/java/com/transferwise/tasks/dao/MySqlTaskDao.java [hunk 1]
```diff
@@ -502,20 +497,52 @@ public DeleteFinishedOldTasksResult deleteOldTasks(TaskStatus taskStatus, Durati
...
     for (int b = questionBuckets.length - 1; b >= 0; b--) {
-      int bucketSize = questionBuckets[b];
-      while (taskIds.size() - processedIdsCount >= bucketSize) {
+      var bucketSize = questionBuckets[b];
+
+      while (taskVersionIds.size() - processedIdsCount >= bucketSize) {
+        final int currentProcessedIdsCount = processedIdsCount;
+
+        if (tasksProperties.isParanoidTasksCleaning()) {
+          var tasksDeleteLockSql = sqlCache.computeIfAbsent(new CacheKey(LOCK_TASKS_FOR_DELETE_SQL, b),
```
Comment (onukristo) [inline tw-tasks-core/src/main/java/com/transferwise/tasks/dao/MySqlTaskDao.java:506]: Basically lock records with SELECT FOR UPDATE.
Then take only those which versions have not changed between earlier select and this.

File: tw-tasks-core/src/main/java/com/transferwise/tasks/dao/MySqlTaskDao.java [hunk 2]
```diff
@@ -562,86 +589,12 @@ public DeleteFinishedOldTasksResult deleteOldTasks(TaskStatus taskStatus, Durati
...
-
-      @Override
-      public int getBatchSize() {
-        return taskVersionIds.size();
-      }
-    });
-
-    List<Object> deletedIds = new ArrayList<>();
-    for (int i = 0; i < tasksDeleteResult.length; i++) {
-      if (tasksDeleteResult[i] > 0) {
```
Comment (onukristo) [inline tw-tasks-core/src/main/java/com/transferwise/tasks/dao/MySqlTaskDao.java:611]: This does not work anymore with latest mariadb driver. It does not give back individual affected rows count.

File: tw-tasks-core/src/test/resources/logback.xml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+<configuration>
+  <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
+    <encoder>
+      <pattern>%d{HH:mm:ss.SSS} [%thread] %-5level %logger{50} - %msg%n</pattern>
+    </encoder>
+  </appender>
+
+  <root level="info">
+    <appender-ref ref="STDOUT"/>
+    <appender-ref ref="TEST"/>
```
Comment (tw-peeterkarolin) [inline tw-tasks-core/src/test/resources/logback.xml:10]: TEST appender doesn't seem to exist.

File: tw-tasks-core/src/main/java/com/transferwise/tasks/triggering/KafkaTasksExecutionTriggerer.java [hunk 4]
```diff
@@ -369,49 +371,79 @@ void releaseCompletedOffset(ConsumerBucket consumerBucket, TopicPartition topicP
...
+    try {
+      if (log.isDebugEnabled()) {
+        log.debug("Sync-committing bucket '" + bucketId + "' offsets to Kafka: " + offsetsToCommit.entrySet().stream()
+            .map(e -> e.getKey() + ":" + e.getValue().offset()).collect(Collectors.joining(", ")));
+      }
+      consumerBucket.getKafkaConsumer().commitSync(offsetsToCommit);
+    } catch (Throwable t) {
+      registerCommitException(bucketId, t);
+    } finally {
+      coreMetricsTemplate.registerKafkaTasksExecutionTriggererCommit(bucketId, true, success);
```
Comment (tw-peeterkarolin) [inline tw-tasks-core/src/main/java/com/transferwise/tasks/triggering/KafkaTasksExecutionTriggerer.java:391]: Where is the success boolean set to true?

File: tw-tasks-core/src/test/resources/logback.xml (hunk 3 above)
Comment (onukristo) [reply tw-tasks-core/src/test/resources/logback.xml]: Fixed

File: tw-tasks-core/src/main/java/com/transferwise/tasks/triggering/KafkaTasksExecutionTriggerer.java (hunk 4 above)
Comment (onukristo) [reply tw-tasks-core/src/main/java/com/transferwise/tasks/triggering/KafkaTasksExecutionTriggerer.java:392]: Fixed.

**Review by tw-peeterkarolin — APPROVED** [review]

**Pull request merged**